File: open-seo-main/src/db/dual-write.ts
/**
 * Dual-write module for database consolidation.
 * Phase 67-03: Cutover
 *
 * Implements shadow write pattern for zero-downtime migration:
 *   - Primary write (blocking): writes to current database
 *   - Shadow write (fire-and-forget): writes to tevero consolidated database
 *
 * Environment variables:
 *   - SHADOW_WRITE_ENABLED: Set to 'true' to enable shadow writes
 *   - TEVERO_DATABASE_URL: Connection string for consolidated database
 *
 * Requirements:
 *   - HIGH-DB-003: Zero-downtime cutover
 *   - MED-DB-007: Feature flag controlled migration
 */

import { drizzle } from "drizzle-orm/node-postgres";
import pg from "pg";
import { eq } from "drizzle-orm";
import { sharedClients, type SharedClientInsert } from "./schema/shared-clients";

/**
 * Shadow write feature flag.
 * When true, writes to tevero database in addition to primary.
 */
const SHADOW_WRITE_ENABLED = process.env.SHADOW_WRITE_ENABLED === "true";

/**
 * Tevero database URL for consolidated database.
 */
const TEVERO_DATABASE_URL = process.env.TEVERO_DATABASE_URL;

/**
 * Cached pool for tevero database connection.
 * Lazily initialized on first shadow write.
 */
let teveroPool: pg.Pool | null = null;

/**
 * Creates or returns cached tevero database connection.
 * Uses a separate pool from primary to avoid connection interference.
 *
 * @returns Drizzle database client for tevero database
 */
export function createTeveroDb() {
  if (!TEVERO_DATABASE_URL) {
    throw new Error(
      "TEVERO_DATABASE_URL environment variable is required for shadow writes"
    );
  }

  if (!teveroPool) {
    teveroPool = new pg.Pool({
      connectionString: TEVERO_DATABASE_URL,
      max: 5, // Smaller pool for shadow writes
      idleTimeoutMillis: 30_000,
      connectionTimeoutMillis: 10_000,
      allowExitOnIdle: true,
      ssl:
        process.env.NODE_ENV === "production"
          ? { rejectUnauthorized: true }
          : false,
    });

    teveroPool.on("error", (err) => {
      console.error("[dual-write] Tevero pool error:", err);
    });
  }

  return drizzle(teveroPool, {
    schema: { sharedClients },
  });
}

/**
 * Closes the tevero database pool.
 * Call during application shutdown.
 */
export async function closeTeveroPool(): Promise<void> {
  if (teveroPool) {
    await teveroPool.end();
    teveroPool = null;
    console.log("[dual-write] Tevero pool closed");
  }
}

/**
 * Client data for dual-write operations.
 */
export type ClientData = SharedClientInsert;

/**
 * Performs a shadow write to tevero database.
 * Fire-and-forget pattern: does not block primary operation.
 *
 * @param operation - The operation type ('insert' or 'update')
 * @param data - The client data to write
 * @param clientId - For updates, the client ID to update
 */
async function shadowWrite(
  operation: "insert" | "update",
  data: ClientData,
  clientId?: string
): Promise<void> {
  if (!SHADOW_WRITE_ENABLED) {
    return;
  }

  if (!TEVERO_DATABASE_URL) {
    console.warn("[dual-write] TEVERO_DATABASE_URL not set, skipping shadow write");
    return;
  }

  try {
    const teveroDb = createTeveroDb();

    if (operation === "insert") {
      await teveroDb.insert(sharedClients).values(data);
      console.log("[dual-write] Shadow insert successful");
    } else if (operation === "update" && clientId) {
      await teveroDb
        .update(sharedClients)
        .set(data)
        .where(eq(sharedClients.id, clientId));
      console.log("[dual-write] Shadow update successful");
    }
  } catch (err) {
    // Fire-and-forget: log error but don't throw
    console.error("[dual-write] Shadow write failed:", err);
  }
}

/**
 * Dual-write insert: creates client in primary DB with shadow write to tevero.
 *
 * @param primaryDb - Primary database client
 * @param data - Client data to insert
 * @returns The inserted client from primary database
 */
export async function dualWriteClientInsert<T>(
  primaryDb: { insert: (table: typeof sharedClients) => { values: (data: ClientData) => { returning: () => Promise<T[]> } } },
  data: ClientData
): Promise<T[]> {
  // Primary write (blocking)
  const result = await primaryDb
    .insert(sharedClients)
    .values(data)
    .returning();

  // Shadow write (fire-and-forget)
  shadowWrite("insert", data).catch((err) =>
    console.error("[dual-write] Shadow write failed:", err)
  );

  return result;
}

/**
 * Dual-write update: updates client in primary DB with shadow write to tevero.
 *
 * @param primaryDb - Primary database client
 * @param clientId - The client ID to update
 * @param data - Client data to update
 * @returns The updated client from primary database
 */
export async function dualWriteClientUpdate<T>(
  primaryDb: {
    update: (table: typeof sharedClients) => {
      set: (data: Partial<ClientData>) => {
        where: (condition: ReturnType<typeof eq>) => { returning: () => Promise<T[]> };
      };
    };
  },
  clientId: string,
  data: Partial<ClientData>
): Promise<T[]> {
  // Primary write (blocking)
  const result = await primaryDb
    .update(sharedClients)
    .set(data)
    .where(eq(sharedClients.id, clientId))
    .returning();

  // Shadow write (fire-and-forget)
  shadowWrite("update", data as ClientData, clientId).catch((err) =>
    console.error("[dual-write] Shadow write failed:", err)
  );

  return result;
}

/**
 * Simple dual-write client function for basic operations.
 * For backwards compatibility with plan specification.
 *
 * @param operation - 'insert' or 'update'
 * @param data - Client data
 * @param primaryDb - Primary database client
 * @param clientId - For updates, the client ID
 * @returns Result from primary write
 */
export async function dualWriteClient(
  operation: "insert" | "update",
  data: ClientData,
  primaryDb: ReturnType<typeof drizzle>,
  clientId?: string
) {
  if (operation === "insert") {
    // Primary write (blocking)
    const result = await primaryDb.insert(sharedClients).values(data).returning();

    // Shadow write (fire-and-forget)
    if (SHADOW_WRITE_ENABLED) {
      shadowWrite("insert", data).catch((err) =>
        console.error("[dual-write] Shadow write failed:", err)
      );
    }

    return result;
  } else if (operation === "update" && clientId) {
    // Primary write (blocking)
    const result = await primaryDb
      .update(sharedClients)
      .set(data)
      .where(eq(sharedClients.id, clientId))
      .returning();

    // Shadow write (fire-and-forget)
    if (SHADOW_WRITE_ENABLED) {
      shadowWrite("update", data, clientId).catch((err) =>
        console.error("[dual-write] Shadow write failed:", err)
      );
    }

    return result;
  }

  throw new Error(`Invalid operation: ${operation}`);
}

/**
 * Check if shadow write is enabled.
 */
export function isShadowWriteEnabled(): boolean {
  return SHADOW_WRITE_ENABLED;
}

File: AI-Writer/backend/services/dual_write.py
"""
Dual-write module for database consolidation.
Phase 67-03: Cutover

Implements shadow write pattern for zero-downtime migration:
  - Primary write (blocking): writes to current database
  - Shadow write (fire-and-forget): writes to tevero consolidated database

Environment variables:
  - SHADOW_WRITE_ENABLED: Set to 'true' to enable shadow writes
  - TEVERO_DATABASE_URL: Connection string for consolidated database

Requirements:
  - HIGH-DB-003: Zero-downtime cutover
  - MED-DB-007: Feature flag controlled migration
"""

import asyncio
import os
from typing import Any, Dict, Optional
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine, AsyncEngine
from sqlalchemy.exc import SQLAlchemyError
from loguru import logger

# Feature flag for shadow writes
SHADOW_WRITE_ENABLED = os.getenv("SHADOW_WRITE_ENABLED", "false").lower() == "true"

# Tevero database URL for consolidated database
TEVERO_DATABASE_URL = os.getenv("TEVERO_DATABASE_URL")

# Cached async engine for tevero database
_tevero_engine: Optional[AsyncEngine] = None


def _get_async_url(url: str) -> str:
    """Convert sync database URL to async URL."""
    if url.startswith("postgresql://"):
        return url.replace("postgresql://", "postgresql+asyncpg://", 1)
    elif url.startswith("postgres://"):
        return url.replace("postgres://", "postgresql+asyncpg://", 1)
    return url


def get_tevero_engine() -> Optional[AsyncEngine]:
    """
    Get or create async engine for tevero database.

    Returns None if TEVERO_DATABASE_URL is not set.
    """
    global _tevero_engine

    if not TEVERO_DATABASE_URL:
        return None

    if _tevero_engine is None:
        async_url = _get_async_url(TEVERO_DATABASE_URL)
        _tevero_engine = create_async_engine(
            async_url,
            pool_size=5,  # Smaller pool for shadow writes
            max_overflow=5,
            pool_timeout=10,
            pool_recycle=1800,
            echo=False,
        )
        logger.info("[dual-write] Tevero async engine created")

    return _tevero_engine


async def close_tevero_engine() -> None:
    """
    Close the tevero database engine.
    Call during application shutdown.
    """
    global _tevero_engine

    if _tevero_engine is not None:
        await _tevero_engine.dispose()
        _tevero_engine = None
        logger.info("[dual-write] Tevero engine closed")


async def shadow_write_client(client_data: Dict[str, Any]) -> None:
    """
    Fire-and-forget shadow write to tevero database.

    Writes client data to shared_clients table in consolidated database.
    Does not block the primary operation - errors are logged and ignored.

    Args:
        client_data: Dictionary with client fields matching shared_clients schema
    """
    if not SHADOW_WRITE_ENABLED:
        return

    if not TEVERO_DATABASE_URL:
        logger.warning("[dual-write] TEVERO_DATABASE_URL not set, skipping shadow write")
        return

    try:
        engine = get_tevero_engine()
        if engine is None:
            return

        # Build insert statement
        columns = list(client_data.keys())
        placeholders = [f":{col}" for col in columns]

        insert_sql = text(f"""
            INSERT INTO shared_clients ({", ".join(columns)})
            VALUES ({", ".join(placeholders)})
            ON CONFLICT (id) DO UPDATE SET
            {", ".join(f"{col} = EXCLUDED.{col}" for col in columns if col != "id")}
        """)

        async with engine.begin() as conn:
            await conn.execute(insert_sql, client_data)

        logger.info("[dual-write] Shadow write client successful")

    except SQLAlchemyError as e:
        # Fire-and-forget: log error but don't raise
        logger.error(f"[dual-write] Shadow write failed: {e}")
    except Exception as e:
        logger.error(f"[dual-write] Shadow write unexpected error: {e}")


async def shadow_update_client(client_id: str, update_data: Dict[str, Any]) -> None:
    """
    Fire-and-forget shadow update to tevero database.

    Updates client data in shared_clients table in consolidated database.
    Does not block the primary operation - errors are logged and ignored.

    Args:
        client_id: The client UUID to update
        update_data: Dictionary with fields to update
    """
    if not SHADOW_WRITE_ENABLED:
        return

    if not TEVERO_DATABASE_URL:
        logger.warning("[dual-write] TEVERO_DATABASE_URL not set, skipping shadow update")
        return

    try:
        engine = get_tevero_engine()
        if engine is None:
            return

        # Build update statement
        set_clauses = [f"{col} = :{col}" for col in update_data.keys()]
        update_data["id"] = client_id

        update_sql = text(f"""
            UPDATE shared_clients
            SET {", ".join(set_clauses)}, updated_at = NOW()
            WHERE id = :id
        """)

        async with engine.begin() as conn:
            await conn.execute(update_sql, update_data)

        logger.info("[dual-write] Shadow update client successful")

    except SQLAlchemyError as e:
        logger.error(f"[dual-write] Shadow update failed: {e}")
    except Exception as e:
        logger.error(f"[dual-write] Shadow update unexpected error: {e}")


def fire_and_forget_shadow_write(client_data: Dict[str, Any]) -> None:
    """
    Synchronous wrapper for fire-and-forget shadow write.

    Creates a background task that doesn't block the current thread.
    Use this from sync code paths.

    Args:
        client_data: Dictionary with client fields matching shared_clients schema
    """
    if not SHADOW_WRITE_ENABLED:
        return

    try:
        loop = asyncio.get_event_loop()
        if loop.is_running():
            # Schedule as background task
            asyncio.create_task(shadow_write_client(client_data))
        else:
            # Run synchronously if no event loop
            loop.run_until_complete(shadow_write_client(client_data))
    except RuntimeError:
        # No event loop - create one for this operation
        asyncio.run(shadow_write_client(client_data))


def fire_and_forget_shadow_update(client_id: str, update_data: Dict[str, Any]) -> None:
    """
    Synchronous wrapper for fire-and-forget shadow update.

    Creates a background task that doesn't block the current thread.
    Use this from sync code paths.

    Args:
        client_id: The client UUID to update
        update_data: Dictionary with fields to update
    """
    if not SHADOW_WRITE_ENABLED:
        return

    try:
        loop = asyncio.get_event_loop()
        if loop.is_running():
            asyncio.create_task(shadow_update_client(client_id, update_data))
        else:
            loop.run_until_complete(shadow_update_client(client_id, update_data))
    except RuntimeError:
        asyncio.run(shadow_update_client(client_id, update_data))


def is_shadow_write_enabled() -> bool:
    """Check if shadow write is enabled."""
    return SHADOW_WRITE_ENABLED
